Add `HammingWeightCompute` gate using 3-to-2 Gidney adders

tanujkhattar opened the pull request:
Follows construction described on Page-4 of https://arxiv.org/abs/1709.06648

cc @ncrubin

fdmalone commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
 
+@frozen
+class HammingWeightCompute(GateWithRegisters):
+    r"""A gate to compute the hamming weight of n-bit register in a new log_{n} bit register.
```
nit: "an n-bit register"

fdmalone commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+class HammingWeightCompute(GateWithRegisters):
+    r"""A gate to compute the hamming weight of n-bit register in a new log_{n} bit register.
+
+    Implements $U|x\rangle |0\rangle \rightarrow |x\rangle|hamming_weight(x)\rangle$
```
probably want some \mathrm's around the hamming_weight

fdmalone replied:
And also check that the `_` renders properly sometimes this has a weird effect

tanujkhattar replied:
Done

fdmalone commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+     - out: A log2(bitize)-sized LEFT/RIGHT register depending on whether the gate adjoint or not.
+
+    References:
+        [Halving the cost of quantum addition](https://arxiv.org/abs/1709.06648)
```
add a page number ~ rough location if relevant.

fdmalone commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        )
+
+    @cached_property
+    def _junk_sizes(self):
```
return type -> List[int]?

tanujkhattar replied:
removed `junk_sizes`

fdmalone commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return ret
+
+    def short_name(self) -> str:
+        return "out = x.bit_count()"
```
nice

fdmalone commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+    def decompose_from_registers(
+        self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
+    ) -> cirq.OP_TREE:
+        x: List[cirq.Qid] = [*quregs['x'][::-1]]
```
Why are we reversing the qubit order? worth a note I think.

tanujkhattar replied:
Added a note

fdmalone reviewed (changes requested):
LGTM % nits

ncrubin commented on qualtran/bloqs/arithmetic.py:
```diff
@@ -578,6 +578,109 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+    r"""A gate to compute the hamming weight of n-bit register in a new log_{n} bit register.
+
+    Implements $U|x\rangle |0\rangle \rightarrow |x\rangle|hamming_weight(x)\rangle$
+    using $4nT$ gates. Uncomputation requires 0 T-gates.
```
Ian's paper and the Hubbard paper both say n-1 toffoli  But Craig's paper says n Toffoli.  Is there an accounting error? or a reporting error?

tanujkhattar commented:
@fdmalone Addressed all comments, improved the bounds on Toffoli's and ancilla's in the docstrings and simplified computation of `junk_bitsize`. I think this is good to go now.
